https://www.youtube.com/watch?v=TLf90ipMzfE

In [ ]:
# connect your Google Drive
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

In [ ]:
!pip install langchain
!pip install openai
!pip install faiss-cpu

!pip install pypdf
!pip install unstructured
!pip install pytesseract

In [ ]:
# https://platform.openai.com/account/billing/overview
import os
os.environ["OPENAI_API_KEY"] = ""

from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.text_splitter import CharacterTextSplitter
from langchain.vectorstores import ElasticVectorSearch, Pinecone, Weaviate, FAISS

from IPython.display import display, Markdown
from langchain.docstore.document import Document

## URL
- https://python.langchain.com/en/latest/modules/indexes/document_loaders/examples/url.html
- https://python.langchain.com/en/latest/modules/indexes/document_loaders/examples/web_base.html

In [ ]:
urls = ["https://www.mizuhobank.co.jp/fintec/guide/index.html","https://www.mizuhobank.co.jp/fintec/recruit/index.html","https://www.mizuhobank.co.jp/fintec/recruit/environment.html"]

In [ ]:
from langchain.document_loaders import UnstructuredURLLoader
loader = UnstructuredURLLoader(urls=urls)
raw_text = loader.load()
raw_text = [s.page_content for s in raw_text]
print(len(raw_text))

3


In [ ]:
from langchain.document_loaders import WebBaseLoader
loader = WebBaseLoader(urls)
raw_text = loader.load()
raw_text = [s.page_content for s in raw_text]
print(len(raw_text))

3


## PDF
- https://python.langchain.com/en/latest/modules/indexes/document_loaders/examples/pdf.html

In [ ]:
from langchain.document_loaders import PyPDFLoader
reader = PyPDFLoader("/content/drive/My Drive/data/Supply chain logistics with quantum and classical annealing algorithms.pdf")
reader = PyPDFLoader("/content/drive/My Drive/data/fg_fy.pdf")
# ページごと（ページ最後の文章など前後関係がつけられなくなってしまう）
raw_text = reader.load_and_split()
# テキスト連結
raw_text = ''.join([s.page_content for s in raw_text])

## YouTube transcripts
- https://python.langchain.com/en/latest/modules/indexes/document_loaders/examples/youtube_transcript.html

In [ ]:
!pip install youtube-transcript-api
!pip install pytube

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.2/57.2 kB 7.9 MB/s eta 0:00:00


In [ ]:
from langchain.document_loaders import YoutubeLoader
loader = YoutubeLoader.from_youtube_url("https://www.youtube.com/watch?v=Ix9WIZpArm0&t=463s", add_video_info=False)
loader.load()

[Document(page_content="is mayor from chartered data and today we're going to be talking about how do you chat with multiple massive PDF documents across multiple files say in this case we're looking at Tesla we're looking at annual reports for 2022 2021 2020 and each year the PDF files are huge like here is 2022 is 449 and then 20 20 21 is around 300 and you add it up you're looking about a thousand pages of PDFs and it just goes on and on and on and just Financial reports and all that kind of stuff extremely tedious to go through and read and what we want is a situation like this where you go Warren Buffett asking you what you like to analyze but we can ask a question about specific PDF so what are the or what were the risk factors for Tesla 2022. let's say I start searching and it's going to give us a response so let's talk about cyclical industry inflationary pressures increasing interest rates the pandemic and what we want to do now is cross check because it's going to give us a r

## Gutenberg
- https://python.langchain.com/en/latest/modules/indexes/document_loaders/examples/gutenberg.html

In [ ]:
from langchain.document_loaders import GutenbergLoader
loader = GutenbergLoader('https://www.gutenberg.org/cache/epub/69972/pg69972.txt')
data = loader.load()
print(data[0].page_content[:300])
print(data[0].metadata)

The Project Gutenberg eBook of The changed brides, by Emma Dorothy


Eliza Nevitte Southworth





This eBook is for the use of anyone anywhere in the United States and


most other parts of the world at no cost and with almost no restrictions


whatsoever. You may copy it, give it away or re-u
{'source': 'https://www.gutenberg.org/cache/epub/69972/pg69972.txt'}


## Hacker News
- https://python.langchain.com/en/latest/modules/indexes/document_loaders/examples/hacker_news.html

In [ ]:
from langchain.document_loaders import HNLoader
loader = HNLoader("https://news.ycombinator.com/item?id=34817881")
data = loader.load()
print(data[0].page_content[:300])
print(data[0].metadata)

delta_p_delta_x 78 days ago  
             | next [–] 

Astrophysical and cosmological simulations are often insightful. They're also very cross-disciplinary; besides the obvious astrophysics, there's networking and sysadmin, parallel computing and algorithm theory (so that the simulation programs a
{'source': 'https://news.ycombinator.com/item?id=34817881', 'title': 'What Lights the Universe’s Standard Candles?'}


## Arxiv
- https://python.langchain.com/en/latest/modules/indexes/document_loaders/examples/arxiv.html

In [ ]:
!pip install arxiv
!pip install pymupdf

In [ ]:
from langchain.document_loaders import ArxivLoader
docs = ArxivLoader(query="strawberry", load_max_docs=2).load()
len(docs)

2

In [ ]:
docs[0].metadata

{'Published': '2023-01-10',
 'Title': 'Autonomous Strawberry Picking Robotic System (Robofruit)',
 'Authors': 'Soran Parsa, Bappaditya Debnath, Muhammad Arshad Khan, Amir Ghalamzan E.',
 'Summary': 'Challenges in strawberry picking made selective harvesting robotic technology\ndemanding. However, selective harvesting of strawberries is complicated forming\na few scientific research questions. Most available solutions only deal with a\nspecific picking scenario, e.g., picking only a single variety of fruit in\nisolation. Nonetheless, most economically viable (e.g. high-yielding and/or\ndisease-resistant) varieties of strawberry are grown in dense clusters. The\ncurrent perception technology in such use cases is inefficient. In this work,\nwe developed a novel system capable of harvesting strawberries with several\nunique features. The features allow the system to deal with very complex\npicking scenarios, e.g. dense clusters. Our concept of a modular system makes\nour system reconfigura

In [ ]:
docs[0].metadata['Summary']

'Challenges in strawberry picking made selective harvesting robotic technology\ndemanding. However, selective harvesting of strawberries is complicated forming\na few scientific research questions. Most available solutions only deal with a\nspecific picking scenario, e.g., picking only a single variety of fruit in\nisolation. Nonetheless, most economically viable (e.g. high-yielding and/or\ndisease-resistant) varieties of strawberry are grown in dense clusters. The\ncurrent perception technology in such use cases is inefficient. In this work,\nwe developed a novel system capable of harvesting strawberries with several\nunique features. The features allow the system to deal with very complex\npicking scenarios, e.g. dense clusters. Our concept of a modular system makes\nour system reconfigurable to adapt to different picking scenarios. We designed,\nmanufactured, and tested a picking head with 2.5 DOF (2 independent mechanisms\nand 1 dependent cutting system) capable of removing possibl

In [ ]:
docs[0].page_content

'Autonomous Strawberry Picking Robotic System\nSoran Parsa∗\nLincoln Institute for Agri-food Technology\nUniversity of Lincoln\nLincoln, UK\nsoran.parsa@gmail.com\nBappaditya Debnath\nKings College London\nLondon, UK\nb.debnath2017@gmail.com\nMuhammad Arshad Khan\nLincoln Institute for Agri-food Technology\nUniversity of Lincoln\nLincoln, UK\nMuKhan@lincoln.ac.uk\nAmir Ghalamzan E.∗†; ‡\nLincoln Institute for Agri-food Technology\nUniversity of Lincoln\nLincoln, UK\naghalamzanesfahani@lincoln.ac.uk\nAbstract\nChallenges in strawberry picking made selective harvesting robotic technology very demand-\ning. However, the selective harvesting of strawberries is a complicated robotic task forming\na few scientific research questions. Most available solutions only deal with a specific picking\nscenario, e.g., picking only a single variety of fruit in isolation. Nonetheless, most econom-\nically viable (e.g. high-yielding and/or disease-resistant) varieties of strawberry are grown\nin dense clus

## 前処理

In [ ]:
import re
code_regex = re.compile('[!"#$%&\'\\\\()*+,-./:;<=>?@[\\]^_`{|}~「」〔〕“”〈〉『』【】＆＊・（）＄＃＠。、？！｀＋￥％…]')
if type(raw_text)==str:
  raw_text = code_regex.sub('', raw_text)
if type(raw_text)==list:
  raw_text = [Document(page_content=code_regex.sub('', s)) for s in raw_text]

## 文章のチャンク化

In [ ]:
# 情報検索の際に、トークンのサイズ制限に引っかからないように、読んだテキストを小さな塊に分割。1000トークン、チャンク間は200だけ重なるように分割
text_splitter = CharacterTextSplitter(        
    separator = "\n",
    chunk_size = 400,
    chunk_overlap  = 50,
    length_function = len,
)
if type(raw_text)==str:
  texts = text_splitter.split_text(raw_text)
if type(raw_text)==list:
  texts = text_splitter.split_documents(raw_text)
  texts = [s.page_content for s in texts]

In [ ]:
len(texts)

751

In [ ]:
texts[0]

'有価証券報告書\n１本書はEDINETElectronic Disclosure for Investors NETwork\nシステムを利用して金融庁に提出した有価証券報告書の記載事項を\n出力印刷したものであります\n２ 本書には上記の方法により提出した有価証券報告書に添付された\n監査報告書及び上記の有価証券報告書と併せて提出した内部統制\n報告書確認書を末尾に綴じ込んでおります \n株式会社みずほフィナンシャルグループ \nE03615 事 業 年 度\n第 20 期自\u30002021年４月１日\n至\u30002022年３月31日頁\n表紙   1\n第一部 企業情報  2\n第１企業の概況  2\n１．主要な経営指標等の推移  2\n２．沿革  5\n３．事業の内容  7\n４．関係会社の状況  10\n５．従業員の状況  16\n第２事業の状況  17\n１．経営方針経営環境及び対処すべき課題等  17\n２．事業等のリスク  24'

In [ ]:
texts[-1]

'接的に関連する損益科目に至る業務プロセスを評価の対象としたさらに選定した重要な事業拠点にかかわらず\nそれ以外の事業拠点をも含めた範囲について重要な虚偽記載の発生可能性が高く見積りや予測を伴う重要な勘定\n科目に係る業務プロセスやリスクが大きい取引を行っている事業又は業務に係る業務プロセスを財務報告への影響を\n勘案して重要性の大きい業務プロセスとして評価対象に追加している\n３評価結果に関する事項\n\u3000上記の評価の結果当事業年度末日時点において当社の財務報告に係る内部統制は有効であると判断した\n４付記事項\n\u3000該当事項なし\n５特記事項\n\u3000該当事項なし'

## 文章の埋め込みと検索

In [ ]:
# Download embeddings from OpenAI
embeddings = OpenAIEmbeddings()

In [ ]:
docsearch = FAISS.from_texts(texts, embeddings)

## LangChain

In [ ]:
from langchain.chains.question_answering import load_qa_chain
from langchain.chains.qa_with_sources import load_qa_with_sources_chain

from langchain.llms import OpenAI
from langchain.chat_models import ChatOpenAI

In [ ]:
llm = ChatOpenAI(temperature=0, model_name="gpt-3.5-turbo")
# llm = OpenAI(temperature=0)

chain = load_qa_chain(llm, chain_type="stuff")
# chain = load_qa_with_sources_chain(llm, chain_type="stuff")

In [ ]:
query = "労働環境について教えて"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
output_chain = chain.run(input_documents=docs, question=query)
display(Markdown(output_chain))

[0.35154343, 0.35436386, 0.3575907, 0.35905534]


申し訳ありませんが、与えられた文脈からは具体的な労働環境に関する情報は得られません。文脈は、当社グループが環境社会課題に対処するために取り組んでいることや、世界的な気候変動対策の進展について説明しています。

In [ ]:
query = "この文章のキーワードをMarkdown記法で抽出して"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
output_chain = chain.run(input_documents=docs, question=query)
display(Markdown(output_chain))

[0.44249836, 0.45967847, 0.46544528, 0.46787745]


# キーワード抽出

- 温室効果ガス排出量
- カーボンニュートラル
- サステナビリティ
- 顧客ニーズ
- 金融グループ
- 経済情勢
- 新型コロナウイルス
- インフレ圧力
- 商品資源価格
- 経常収益
- コンプライアンス
- 外為法令
- アンチマネーロンダリング
- テロ資金供与対策
- 反社会的勢力

In [ ]:
query = "事業内容・サービスに関するキーワードをMarkdown記法で抽出して"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
output_chain = chain.run(input_documents=docs, question=query)
display(Markdown(output_chain))

In [ ]:
query = "事業内容について要約して"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
query = "事業戦略について要約して"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
query = "リスクについて要約して"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
query = "経営課題について要約して"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
query = "経営課題とその課題に対する事業戦略について要約して"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
query = "Who is the author of this paper?"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
query = "Summarize this paper."
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
query = "Extract Abstract."
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
query = "What is the title of this paper?"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
query = "Summarize the main points of this paper in about three bullet points."
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

In [ ]:
text = "Noisy intermediate-scale quantum (NISQ) hardware is almost universally incompatible with fullscale optimization problems of practical importance which can have many variables and unwieldy objective functions. As a consequence, there is a growing body of literature that tests quantum algorithms on miniaturized versions of problems that arise in an operations research setting. Rather than taking this approach, we investigate a problem of substantial commercial value, multi-truck vehicle routing for supply chain logistics, at the scale used by a corporation in their operations. Such a problem is too complex to be fully embedded on any near-term quantum hardware or simulator; we avoid confronting this challenge by taking a hybrid workflow approach: we iteratively assign routes for trucks by generating a new binary optimization problem instance one truck at a time. Each instance has  2500 quadratic binary variables, putting it in a range that is feasible for NISQ quantum computing, especially quantum annealing hardware. We test our methods using simulated annealing and the D-Wave Hybrid solver as a place-holder in wait of quantum hardware developments. After feeding the vehicle routes suggested by these runs into a highly realistic classical supply chain simulation, we find excellent performance for the full supply chain. Our work gives a set of techniques that can be adopted in contexts beyond vehicle routing to apply NISQ devices in a hybrid fashion to large-scale problems of commercial interest."
query = f"Translate the following sentence into Japanese.\n {text}"
output = docsearch.similarity_search_with_score(query)
docs = [s[0] for s in output]
scores = [s[1] for s in output]
print(scores)
chain.run(input_documents=docs, question=query)

## ChatGPT arXiv

In [ ]:
# In[]
# ref
"""
https://zenn.dev/team_zenn/articles/117424abb5605b
https://zenn.dev/ozushi/articles/ebe3f47bf50a86
https://developers.notion.com/reference/post-page
https://info.arxiv.org/help/api/user-manual.html
https://api.semanticscholar.org/api-docs/graph
"""
import requests
import time
from datetime import datetime, timezone
    
from scipy.stats import rankdata
import numpy as np
    
import openai
import arxiv

# 現在時刻
now = datetime.now(timezone.utc)

#arXiv Settings
ARXIV_NUM_PAPERS = 500
RESULT_NUM_PAPERS = 10

# タイトルキーワード:ti
ARXIV_QUERY = 'ti:%22 reciprocal %22 AND ti:%22 recommend %22'
ARXIV_QUERY = 'ti:%22 strawberry %22'
ARXIV_QUERY = 'ti:%22 tomato %22'

# すべて:all
ARXIV_QUERY = 'all:%22 reciprocal %22 AND all:%22 recommend %22' # reciprocal AND recommendで検索
ARXIV_QUERY = 'all:cs.CV' # コンピュータービジョンとパターン認識
ARXIV_QUERY = 'all:stat.ML' # 機械学習
ARXIV_QUERY = 'all:q-fin.MF' # 数学ファイナンス
ARXIV_QUERY = 'all:q-fin.PM' # ポートフォリオマネジメント
ARXIV_QUERY = 'all:q-fin.PR' # 証券のプライシング
ARXIV_QUERY = 'all:q-fin.RM' # リスク管理
ARXIV_QUERY = 'all:q-fin.TR' # Trading and Market Microstructure

# カテゴリ(いまいち？):cat
# ARXIV_QUERY = 'cat:cs.AI' # 人工知能


#OpenAI Settings
openai.api_key = ""
OPENAI_PROMPT = """与えられた論文の要点を3点のみでまとめ、以下のフォーマットで日本語で出力してください。```
Titleの日本語訳
・要点1
・要点2
・要点3
```"""

#Notion Settings
NOTION_API_KEY = ""
DATABASE_ID = ""
NOTION_API_URL = 'https://api.notion.com/v1/pages'

headers = {
    "Accept": "application/json",
    "Notion-Version": "2022-06-28",
    "Content-Type": "application/json",
    "Authorization": "Bearer " + NOTION_API_KEY
}

payload = {
    "parent": {
        "database_id": DATABASE_ID
    },
    "properties": {
        "Title": {
            "title": [
            {
                "text": {
                "content": ""
                }
            }
            ],
        },
        "タイトル": {
            "rich_text": [
            {
                "text": {
                "content": ""
                }
            }
            ],
        },
        "Detail": {
            "rich_text": [
            {
                "text": {
                "content": ""
                }
            }
            ],
        },
        "URL":{
            "url": ""
        },
        "Citation": {
            "number": 0
        },
        "InfluentialCitation": {
            "number": 0
        },
        "Date": {
            "date": {
                "start": ""
                }
            },
        "arXiv query": {
            "rich_text": [
            {
                "text": {
                "content": ARXIV_QUERY
                }
            }
            ],
        },
    },
    "children":[
        {
            "object": 'block',
            "type": 'heading_1',
            "heading_1":{
                "rich_text": [
            {
                "text": {
                "content": "Abstract"
                }
            }
            ],
            }
        },
        {
            "object": 'block',
            "type": 'paragraph',
            "paragraph":{
                "rich_text": [
            {
                "text": {
                "content": ""
                }
            }
            ],
            }
        },
        {
            "object": 'block',
            "type": 'heading_1',
            "heading_1":{
                "rich_text": [
            {
                "text": {
                "content": "要約"
                }
            }
            ],
            }
        },
        {
            "object": 'block',
            "type": 'paragraph',
            "paragraph":{
                "rich_text": [
            {
                "text": {
                "content": ""
                }
            }
            ],
            }
        },
        {
            "object": 'block',
            "type": 'heading_1',
            "heading_1":{
                "rich_text": [
            {
                "text": {
                "content": "TLDR"
                }
            }
            ],
            }
        },
        {
            "object": 'block',
            "type": 'paragraph',
            "paragraph":{
                "rich_text": [
            {
                "text": {
                "content": ""
                }
            }
            ],
            }
        },
        {
            "object": 'block',
            "type": 'heading_1',
            "heading_1":{
                "rich_text": [
            {
                "text": {
                "content": "To do"
                }
            }
            ],
            }
        },
        {
            "object": 'block',
            "type": 'to_do',
            "to_do":{
                "rich_text": [
            {
                "text": {
                "content": ""
                }
            }
            ],
            "checked": False,
            "color": "default",
            }
        }
    ],
}

# In[]
def create_notion_page(result,summary,response_semanticscholar):
    # from arxiv API
    payload['properties']['Title']['title'][0]['text']['content'] = result.title
    payload['properties']['URL']['url'] = result.entry_id
    payload['children'][1]['paragraph']['rich_text'][0]['text']['content'] = result.summary
    payload['properties']['Date']['date']['start'] = result.published.strftime("%Y-%m-%d %H:%M:%S")
    
    # from OpenAI API
    title_jp, *body = summary.split('\n')
    body = '\n'.join(body)
    
    payload['properties']['タイトル']['rich_text'][0]['text']['content'] = title_jp
    payload['properties']['Detail']['rich_text'][0]['text']['content'] = body
    # payload['children'][3]['paragraph']['rich_text'][0]['text']['content'] = abstract_jp
    
    # from Semantic Scholar Academic Graph API
    if not 'error' in response_semanticscholar:
        payload['properties']['Citation']['number'] = response_semanticscholar['citationCount']
        payload['properties']['InfluentialCitation']['number'] = response_semanticscholar['influentialCitationCount']
        if not response_semanticscholar['tldr'] is None:
            payload['children'][5]['paragraph']['rich_text'][0]['text']['content'] = response_semanticscholar['tldr']['text']
        else:
            payload['children'][5]['paragraph']['rich_text'][0]['text']['content'] = ""
            
    else:
        payload['properties']['Citation']['number'] = 0
        payload['properties']['InfluentialCitation']['number'] = 0
        payload['children'][5]['paragraph']['rich_text'][0]['text']['content'] = ""
    
    
    response_notion = requests.post(NOTION_API_URL, json=payload, headers=headers)
    return response_notion

def main():
    # arxiv APIで最新の論文情報を取得する
    search = arxiv.Search(
        query=ARXIV_QUERY,  # 検索クエリ
        max_results=ARXIV_NUM_PAPERS,  # 取得する論文数
        sort_by=arxiv.SortCriterion.Relevance,  # 論文を投稿された日付でソートする(SubmittedDate,Relevance,LastUpdatedDate)
        sort_order=arxiv.SortOrder.Descending,  # 新しい論文から順に取得する(Descending,Ascending)
    )
    
    #searchの結果をリストに格納
    results = []
    arxiv_list = []
    for result in search.results():
        results.append(result)
        arxiv_list.append("ARXIV:"+result.entry_id.split("/")[-1].split("v")[0])
    
    # to Semantic Scholar Academic Graph API
    try:
        response_semanticscholar = requests.post(
        'https://api.semanticscholar.org/graph/v1/paper/batch',
        params={'fields': 'citationCount,influentialCitationCount,tldr'},
        json={"ids": arxiv_list}
        )
        response_semanticscholar = response_semanticscholar.json()
    except:
        print("Semantic Scholar Academic Graph API Error")

    # 並び替え    
    order_list = []
    for i,s in enumerate(response_semanticscholar):
        if s is None:
            order_list.append(0)
        else:
            # 日数あたりの引用回数
            order_list.append(-s['citationCount']/(now-results[i].published).days) # citationCount,influentialCitationCount
    order_list = rankdata(order_list, method='ordinal')-1
    order_list = sorted(range(len(order_list)), key=lambda k: order_list[k])
    
    results = list(np.array(results)[order_list])
    response_semanticscholar = list(np.array(response_semanticscholar)[order_list])
    
    # 論文情報をNotion投稿する
    for i,result in enumerate(results[:RESULT_NUM_PAPERS]):
        # API制限のため（3回/60秒）
        start = time.time()
        
        # to OpenAI API
        try:
            # text = f"Title: {result.title}\n Abstract: {result.summary}\n TLDR: {response_semanticscholar[i]['tldr']['text']}"
            text = f"Title: {result.title}\n Abstract: {result.summary}"
            response_chatgpt = openai.ChatCompletion.create(
                        model="gpt-3.5-turbo",
                        messages=[
                            {'role': 'system', 'content': OPENAI_PROMPT},
                            {'role': 'user', 'content': text}
                        ],
                        temperature=0,
                    )
            summary = response_chatgpt['choices'][0]['message']['content']
        except:
            print("OpenAI API Error")
                    
        # to Notion API
        try:
            print(create_notion_page(result,summary,response_semanticscholar[i]))
        except:
            print("Notion API Error")
            
        elapsed_time = time.time()-start
        if elapsed_time<=20:
            print(f"Sleep for {20-elapsed_time} seconds due to API limitation")
            time.sleep(20-elapsed_time)
            
if __name__ == "__main__":
    main()

https://www.youtube.com/watch?v=Sk1HrR6WGkg

0:22:30（３）九州沖縄経済圏スマートフードチェーンプロジェクト成果の事業化と輸出拡大に向けた農研機構の取組み
　　農研機構本部 総括執行役 兼 事業開発部長 田中 健一

0:41:58（４）各課題研究成果の説明
0:42:31 　１）輸送中のかんしょ腐敗問題の対応
　　農研機構九州沖縄農業研究センター 所長 森田 敏
1:00:57 　２）サツマイモ基腐病対策技術
　　農研機構植物防疫研究部門 所長 眞岡 哲夫
1:23:01 　３）質疑応答
1:55:30 　４）イチゴの輸出促進に向けた課題解決と産地拡大
　　農研機構九州沖縄農業研究センター 所長 森田 敏
2:10:45 　５）緑茶新品種｢せいめい｣の産地形成と高品質･安定生産技術
　　農研機構果樹茶業研究部門 所長 生駒 吉識
2:31:57 　６）牛肉輸出拡大に向けた生産基盤強化技術開発
　　農研機構畜産研究部門 所長 三森 眞琴
2:47:17 　７）質疑応答

輸送中のかんしょ腐敗問題の対応
	* 
高温キュアリング：高温度、高湿度によりコルク層を作る、かんしょの腐敗の減少。根切をするとより下がる（原因にもなっている 


サツマイモ基腐病対策技術
	* 
サツマイモの基腐病：平成30年にはじめて発生、沖縄と鹿児島。以降、令和4年には20数件に拡大 ・いちご：輸出、2022年度予想50億円。急増中。恋みのりではがく枯れ、輸送、単収の構造に対応。


イチゴの輸出促進に向けた課題解決と産地拡大
	* 
がく枯れは土壌含水率が多いとなりやすいと比較実験により発見
	* 
輸送は収穫前後の気温、保管温度、梱包で改善。
	* 
増収：CO2の局所施用と換気の自動制御



緑茶新品種｢せいめい｣の産地形成と高品質･安定生産技術
	* 
2020（令和2）年のお茶の栽培（さいばい）面積は、1位が静岡県（しずおかけん）15,200ha（39%）、2位が鹿児島県（かごしまけん）8,360ha（21%）、3位が三重県2,710ha（7%）、4位が京都府1,560ha（4%）となっています。


牛肉輸出拡大に向けた生産基盤強化技術開発
	* 




https://www.youtube.com/watch?v=IXtxQO9dzXE

(1) 開会
0:00:00 開会

(2) 主催者挨拶
0:00:49 農研機構 理事長　久間 和生

(3) 来賓挨拶
0:07:08 北海道十勝総合振興局長　水戸部 裕 様
0:12:24 フードバレーとかち推進協議会会長　米沢 則寿 様

(4) 講演
0:17:14 北海道十勝発スマートフードチェーンプロジェクトの目指すもの
    農研機構 北海道農業研究センター 所長　安東 郁男

北海道、てんさい、小麦、ばれいしょ、生乳、野菜、米、シェア1位
農家人口、農家戸数の減少に伴う大規模・効率化が必要不可欠
・畑作物の終了予測ノウハウ・知見と目指す予測精度（5%）



0:45:35 北海道十勝発スマートフードチェーンプロジェクトの可能性と期待 (仮題)
　株式会社NTTアグリテクノロジー 代表取締役社長　酒井 大雅 様


1:09:56 畑作物のデータ連携を前提とした高度気象情報と生育・収量予測モデル
　農研機構 北海道農業研究センター 寒地畑作研究領域 領域長　奥野 林太郎
・西日本農研、50mメッシュを地形データなどを使いながら推計する技術を持つ
・てんさい、2ヶ月前時点の精緻な予測があると良い→トラックの集荷、配送の効率化



1:40:13 乳業と酪農の振興に向けたスマート酪農フードチェーンの構築
    農研機構 北海道農業研究センター 寒地酪農研究領域 領域長　須藤 賢司
2:14:42 土壌の特性と可給態養分に基づくスマート施肥技術
    国立大学法人帯広畜産大学 グローバルアグロメディシン研究センター教授 (兼任 : 環境農学研究部門植物生産科学分野長)　谷 昌幸 様

(5) 閉会挨拶
2:43:27 農研機構 理事　折戸 文夫


・